src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs

```diff
@@ -219,6 +223,82 @@ public void TestVectorUdf()
             }
         }
 
+        [Fact]
+        public void TestDataFrameVectorUdf()
+        {
+            Func<PrimitiveDataFrameColumn<int>, ArrowStringDataFrameColumn, ArrowStringDataFrameColumn> udf1Func =
+                (ages, names) =>
+                {
+                    StringArray stringArray = (StringArray)ToArrowArray(
+                    Enumerable.Range(0, (int)names.Length)
+                        .Select(i => $"{names[i]} is {ages[i] ?? 0}")
+                        .ToArray());
+                    return ToArrowStringDataFrameColumn(stringArray);
+                };
+
+            // Single UDF.
+            Func<Column, Column, Column> udf1 =
+                ExperimentalDataFrameFunctions.VectorUdf(udf1Func);
+            {
+                Row[] rows = _df.Select(udf1(_df["age"], _df["name"])).Collect().ToArray();
+                Assert.Equal(3, rows.Length);
+                Assert.Equal("Michael is 0", rows[0].GetAs<string>(0));
+                Assert.Equal("Andy is 30", rows[1].GetAs<string>(0));
+                Assert.Equal("Justin is 19", rows[2].GetAs<string>(0));
+            }
+
+            // Chained UDFs.
+            Func<Column, Column> udf2 = ExperimentalDataFrameFunctions.VectorUdf<ArrowStringDataFrameColumn, ArrowStringDataFrameColumn>(
+                (strings) =>
+                {
+                    StringArray stringArray = (StringArray)ToArrowArray(

```

Comment:
So do we need to first change to arrow array then back to dataframe column? No way to go from `ArrowStringDataFrameColumn` to `ArrowStringDataFrameColumn`? cc: @eerhardt

Reply:
I thought we had an `Apply` function that could be used here.
One of the biggest issues is going from the UTF8 arrow string format, to the UTF16 .NET string format, then back to UTF8 arrow string format.

---
In reply to: [387794908](https://github.com/dotnet/spark/pull/277#discussion_r387794908) [](ancestors = 387794908)

Reply:
Yea, it will be really nice if the user doesn't have to go thru Arrow APIs but lives in FxDataFrame APIs.

Reply:
We don't have it in 0.2.0 yet. However, we should release a 0.3.0 soon which will have this API. How about we accept this change for now and I put up a new PR once 0.3.0 is out?

Reply:
That sounds like a good plan to me.

Reply:
Filed https://github.com/dotnet/corefxlab/issues/2860

Reply:
Sounds good to me as well.